### ISSUE TITLE
Audacity 3.2 for Mac Silicon does not find homebrew-installed ffmpeg

### ISSUE THREAD

**Issue by M-D-M** [audacity/audacity#3747]
### Bug description

Summary:

Wanted to use the new version of Audacity for Mac Silicon (thanks for making it!), but unfortunately it does not find the version of ffmpeg I have installed through homebrew.

Further Description:

Installed Mac Silicon version of Audacity via .dmg file.  Selected "do not replace previous version" when prompted via MacOS, because I wanted to keep the previous x86 version handy.

When loading Audacity, it says that ffmpeg libraries aren't loaded, and when trying to direct it to the folder with ffmpeg libraries installed via Homebrew (/opt/homebrew/Cellar/ffmpeg/*/lib), it does not find the right library.

### Steps to reproduce

Download Mac Silicon version of Audacity.

Install using dmg file, and select "do not replace previous version" when prompted via MacOS.

Open Audacity, and get the popup about ffmpeg libraries not found.

### Expected behavior

Audacity finds the ffmpeg libraries.

### Actual behavior

Audacity is saying that "ffmpeg library not found." 

### Audacity Version

current master / alpha / nightly build

### Operating system

macOS

### Additional context

I have ffmpeg installed and in my path via Homebrew -- confirmed with "ffmpeg -version" from a cli.  

It looks like audacity is wanting a file named "ffmpeg.64bit.dylib", but the files in the ffmpeg folder have names like the following:

libavcodec.59.37.100.dylib
libavdevice.59.7.100.dylib
libavfilter.8.44.100.dylib
libavformat.59.27.100.dylib

**Comment by dozzzzer** [comment]
@M-D-M we will look into it, but for now please download and install the FFmpeg package for OS X [from Buanzo's website](https://lame.buanzo.org/).

**Comment by dozzzzer** [comment]
Verified. I don't see the original issue anymore. Now Audacity recognizes FFmpeg installed by Homebrew.

<img width="550" alt="image" src="https://user-images.githubusercontent.com/70063970/193810453-0b0c60df-cd89-4f1f-a299-154b4206e6e7.png">

**Comment by M-D-M** [comment]
Thank you for looking into this so quickly! 

Following up on the ffmpeg package from Buanzo's website -- I already had the x86 libs installed, but could not find any arm libs.

**Comment by M-D-M** [comment]
Was this the key line that solves the issue?

paths.emplace_back(wxT("/opt/homebrew/lib"));

FYI: I tried opening up Audacity for Mac Silicon again, and manually entered this location ("/opt/homebrew/lib") into the preferences, and now the ffmpeg libs are correctly found:

<img width="542" alt="Screen Shot 2022-10-04 at 9 59 02 PM" src="https://user-images.githubusercontent.com/5725775/193963113-637d336c-1ce7-48db-8edd-e1859179bb7a.png">

**Comment by dozzzzer** [comment]
> I already had the x86 libs installed, but could not find any arm libs.

AFAIK, there's one .dmg package for both architectures.

**Comment by LWinterberg** [comment]
(I think the buanzo website was updated after this issue was made)

**Comment by crsib** [comment]
The pull request for this issue is not merged. Why is it closed?

**Comment by BlkPingu** [comment]
I have this issue too with 3.3.3

**Comment by savagekevin** [comment]
same for me as original post, installed Audacity 3.3.3 then installed Homebrew, then home-brew installed ffmpeg, after restarts, Audacity doesn't find it, using the preferences to direct it, Audacity asks for "libavformat.dylib"  but any such files that one points to in the home-brew installation don't seem to please Audacity

**Comment by martingeorgiu** [comment]
I also had the same issue, but I then found out that my local installation of FFmpeg was behaving weirdly. I've reinstalled the FFmpeg from `brew` like described here https://stackoverflow.com/a/77115222/5647161 and then Audacity immediately found it.

**Comment by puppycodes** [comment]
This was also the case for me that some update created an issue with libsodium leading to my ffmpeg woes,  fix was simply to brew upgrade and then it found the lib. It might be helpful to include a verbose error instead of just not finding the library in audacity if thats possible ;)

**Comment by lgarron** [comment]
Anyone else still having issues? I installed both `audacity` (3.5.1) and `ffmpeg` (7.0.1) through Homebrew (4.2.21-43-g020d994) on macOS (14.4.1) but Audacity doesn't find `ffmpeg`.

Here is what I have installed:

```shell
> find /opt -iname "*ffmpeg*"
/opt/homebrew/bin/ffmpeg
/opt/homebrew/var/homebrew/linked/ffmpeg
/opt/homebrew/lib/python3.11/site-packages/yt_dlp/postprocessor/ffmpeg.py
/opt/homebrew/lib/python3.11/site-packages/yt_dlp/postprocessor/__pycache__/ffmpeg.cpython-311.pyc
/opt/homebrew/lib/python3.12/site-packages/yt_dlp/postprocessor/ffmpeg.py
/opt/homebrew/lib/python3.12/site-packages/yt_dlp/postprocessor/__pycache__/ffmpeg.cpython-312.pyc
/opt/homebrew/lib/python3.12/site-packages/torio/utils/__pycache__/ffmpeg_utils.cpython-312.pyc
/opt/homebrew/lib/python3.12/site-packages/torio/utils/ffmpeg_utils.py
/opt/homebrew/lib/python3.12/site-packages/torio/lib/libtorio_ffmpeg5.so
/opt/homebrew/lib/python3.12/site-packages/torio/lib/_torio_ffmpeg4.so
/opt/homebrew/lib/python3.12/site-packages/torio/lib/_torio_ffmpeg5.so
/opt/homebrew/lib/python3.12/site-packages/torio/lib/libtorio_ffmpeg4.so
/opt/homebrew/lib/python3.12/site-packages/torio/lib/_torio_ffmpeg6.so
/opt/homebrew/lib/python3.12/site-packages/torio/lib/libtorio_ffmpeg6.so
/opt/homebrew/lib/python3.12/site-packages/torchaudio/utils/__pycache__/ffmpeg_utils.cpython-312.pyc
/opt/homebrew/lib/python3.12/site-packages/torchaudio/utils/ffmpeg_utils.py
/opt/homebrew/lib/python3.12/site-packages/torchaudio/_backend/ffmpeg.py
/opt/homebrew/lib/python3.12/site-packages/torchaudio/_backend/__pycache__/ffmpeg.cpython-312.pyc
/opt/homebrew/opt/ffmpeg
/opt/homebrew/opt/ffmpeg@7
/opt/homebrew/opt/ffmpeg@6
/opt/homebrew/opt/ffmpeg@5
/opt/homebrew/Cellar/ffmpeg
/opt/homebrew/Cellar/ffmpeg/7.0_1/bin/ffmpeg
/opt/homebrew/Cellar/ffmpeg/7.0_1/.brew/ffmpeg.rb
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/man/man1/ffmpeg-all.1
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/man/man1/ffmpeg-filters.1
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/man/man1/ffmpeg.1
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/man/man1/ffmpeg-codecs.1
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/man/man1/ffmpeg-utils.1
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/man/man1/ffmpeg-resampler.1
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/man/man1/ffmpeg-bitstream-filters.1
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/man/man1/ffmpeg-protocols.1
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/man/man1/ffmpeg-scaler.1
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/man/man1/ffmpeg-devices.1
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/man/man1/ffmpeg-formats.1
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/ffmpeg
/opt/homebrew/Cellar/yt-dlp/2024.4.9_1/libexec/lib/python3.12/site-packages/yt_dlp/postprocessor/ffmpeg.py
/opt/homebrew/Cellar/fish/3.7.1/share/fish/completions/ffmpeg.fish
/opt/homebrew/Cellar/xdg-ninja/0.2.0.2/share/xdg-ninja/programs/ffmpeg.json
/opt/homebrew/share/man/man1/ffmpeg-all.1
/opt/homebrew/share/man/man1/ffmpeg-filters.1
/opt/homebrew/share/man/man1/ffmpeg.1
/opt/homebrew/share/man/man1/ffmpeg-codecs.1
/opt/homebrew/share/man/man1/ffmpeg-utils.1
/opt/homebrew/share/man/man1/ffmpeg-resampler.1
/opt/homebrew/share/man/man1/ffmpeg-bitstream-filters.1
/opt/homebrew/share/man/man1/ffmpeg-protocols.1
/opt/homebrew/share/man/man1/ffmpeg-scaler.1
/opt/homebrew/share/man/man1/ffmpeg-devices.1
/opt/homebrew/share/man/man1/ffmpeg-formats.1
/opt/homebrew/share/ffmpeg
/opt/homebrew/share/fish/completions/ffmpeg.fish
```

```
/opt/homebrew/include/libavformat
/opt/homebrew/lib/pkgconfig/libavformat.pc
/opt/homebrew/lib/libavformat.61.1.100.dylib
/opt/homebrew/lib/libavformat.61.dylib
/opt/homebrew/lib/libavformat.dylib
/opt/homebrew/lib/libavformat.a
/opt/homebrew/Cellar/ffmpeg/7.0_1/include/libavformat
/opt/homebrew/Cellar/ffmpeg/7.0_1/lib/pkgconfig/libavformat.pc
/opt/homebrew/Cellar/ffmpeg/7.0_1/lib/libavformat.61.1.100.dylib
/opt/homebrew/Cellar/ffmpeg/7.0_1/lib/libavformat.61.dylib
/opt/homebrew/Cellar/ffmpeg/7.0_1/lib/libavformat.dylib
/opt/homebrew/Cellar/ffmpeg/7.0_1/lib/libavformat.a
/opt/homebrew/Cellar/ffmpeg/7.0_1/share/man/man3/libavformat.3
/opt/homebrew/share/man/man3/libavformat.3
```

I've tried all of the following paths, and none of them worked:

- `/opt/homebrew/lib`
- `/opt/homebrew/lib/libavformat.dylib`
- `/opt/homebrew/lib/libavformat.61.dylib`
- `/opt/homebrew/lib/libavformat.61.1.100.dylib`
- `/opt/homebrew/Cellar/ffmpeg/7.0_1/lib/` (the folder that the previous `/opt/homebrew/lib/libavformat*.dylib` files are symlinked from)
- `/opt/homebrew/var/homebrew/linked/ffmpeg/lib` (a folder symlink to `/opt/homebrew/Cellar/ffmpeg/7.0_1/lib/`)
- `/opt/homebrew/include/libavformat`

Examples of me trying to tell Audacity to use either the folder or file path where `libavformat.dylib` 100% exists:

![Screenshot 2024-05-07 at 21 24 02](https://github.com/audacity/audacity/assets/248078/027ea65c-d0c4-4b48-8b50-dd27889adcc0)
![Screenshot 2024-05-07 at 21 25 12](https://github.com/audacity/audacity/assets/248078/83c72726-b8af-4add-b20a-f26a98d5036e)

The result remains "Ffmpeg library not found" every single time:
![Screenshot 2024-05-07 at 21 25 34](https://github.com/audacity/audacity/assets/248078/1f9acd07-847a-47b3-8ca3-413d07e12e0d)

… am I missing something?

**Comment by lgarron** [comment]
Okay, I think I figured it out:

`ffmpeg` versions 5 and 6 work.
```
brew install ffmpeg@5
sudo rm -rf "/Library/Application Support/audacity/libs"
sudo ln -s "/opt/homebrew/opt/ffmpeg@5/lib" "/Library/Application Support/audacity/libs"

brew install ffmpeg@6
sudo rm -rf "/Library/Application Support/audacity/libs"
sudo ln -s "/opt/homebrew/opt/ffmpeg@6/lib" "/Library/Application Support/audacity/libs"
```

But version 7 **doesn't** work:

```
brew install ffmpeg # version 7 as of 2024-04-05: https://github.com/Homebrew/homebrew-core/pull/168136
sudo rm -rf "/Library/Application Support/audacity/libs"
sudo ln -s "/opt/homebrew/opt/ffmpeg/lib" "/Library/Application Support/audacity/libs"
```

Homebrew updated to version 7 of `ffmpeg` about a month ago, and I must have picked up that version in an upgrade: https://github.com/Homebrew/homebrew-core/pull/168136

**Comment by lgarron** [comment]
Just to summarize the best workaround for now:

```shell
brew install ffmpeg@6
sudo rm -rf "/Library/Application Support/audacity/libs"
sudo ln -s "/opt/homebrew/opt/ffmpeg@6/lib" "/Library/Application Support/audacity/libs"
```

And once the next version of Audacity is released, you can clean up with:
```shell
brew uninstall ffmpeg@6
sudo rm -rf "/Library/Application Support/audacity/libs"
```

Since `ffmpeg@6` is tagged with a version, it is not linked into the default system paths and will not interfere with `ffmpeg` 7 as the system default for other purposes.

**Comment by lgarron** [comment]
Has anyone seen a regression with this?

I've just installed Audacity and `ffmpeg` on a computer using Homebrew, and Audacity 3.7.5 is unable to find the Homebrew `ffmpeg` (8.0) libs.

It doesn't even work when I do the following:

```shell
sudo mkdir -p "/Library/Application Support/audacity"
sudo ln -s "/opt/homebrew/opt/ffmpeg/lib" "/Library/Application Support/audacity/libs"
```

**Comment by lgarron** [comment]
> I've just installed Audacity and `ffmpeg` on a computer using Homebrew, and Audacity 3.7.5 is unable to find the Homebrew `ffmpeg` (8.0) libs.

Aha, it's because of incompatibility with `ffmpeg 8`. Same dealio as last time, I guess:

```shell
brew install ffmpeg@7
sudo rm -rf "/Library/Application Support/audacity/libs"
sudo mkdir -p "/Library/Application Support/audacity"
sudo ln -s "/opt/homebrew/opt/ffmpeg@7/lib" "/Library/Application Support/audacity/libs"
```

And once https://github.com/audacity/audacity/issues/9252 is fixed, you can clean up with:

```shell
sudo rm -rf "/Library/Application Support/audacity/libs"
brew uninstall ffmpeg@7
```

It sure would be nice if routine system maintenance didn't break Audacity like this, though. 😢

**Comment by jimhigson** [comment]
This works for me:

```sh
# instll ffmpeg v7
brew install ffmpeg@7
```

Then in audacity, set the path for ffmeg to `/opt/homebrew/Cellar/ffmpeg@7/7.1.2/lib/` under:

Settings->Libraries->FFmpeg Library Locate button

**Comment by kryksyh** [comment]
FFmpeg8 support was added to the development branch, and will be available in release 3.7.6.
